src/MongoDB.Driver.Core/Core/NativeLibraryLoader/ILibraryLocator.cs

```diff
@@ -0,0 +1,61 @@
+﻿/* Copyright 2019-present MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.IO;
+using System.Reflection;
+
+namespace MongoDB.Driver.Core.NativeLibraryLoader
+{
+    internal interface ILibraryLocator
+    {
+        string GetLibraryAbsolutePath(SupportedPlatform currentPlatform);
+    }
+
+    internal abstract class LibraryRelativeLocatorBase : ILibraryLocator
+    {
+        public virtual Assembly Assembly => typeof(LibraryRelativeLocatorBase).GetTypeInfo().Assembly;
+
+        public abstract string GetLibraryRelativePath(SupportedPlatform currentPlatform);
+
+        public string GetLibraryAbsolutePath(SupportedPlatform currentPlatform)
+        {
+            var relativePath = GetLibraryRelativePath(currentPlatform);
+            return GetAbsolutePath(relativePath);
+        }
+
+        // private methods
+        private string FindLibraryOrThrow(string basePath, string relativePath)
+        {
+            var fullPath = Path.Combine(basePath, relativePath);
+            if (File.Exists(fullPath))
+            {
+                return fullPath;
+            }
+
+            throw new FileNotFoundException($"Could not find library {fullPath}.");
+        }
+
+        private string GetAbsolutePath(string relativePath)
+        {
+            var codeBase = Assembly.CodeBase;
+            var uri = new Uri(codeBase);
+            var location = uri.AbsolutePath;
+            var basepath = Path.GetDirectoryName(location);
+
+            return FindLibraryOrThrow(basepath, relativePath);
```

Comment:
Refactor to:

    private string GetAbsolutePath(string relativePath)
    {
        var basePath = GetLibraryBasePath();
        return FindLibraryOrThrow(basePath, relativePath);
    }

Reply:
done